### DISCUSSION TITLE
allow to run multiple test in different clients in parallel

### DISCUSSION THREAD

**Discussion by robstoll** [sbt/sbt#6829] [Ideas]
Use case: I have independent tests in two packages and would like to be able to run them in two separate terminals (with `sbt -client`).
i.e. in one terminal I input `~testOnly mypackage1.*` and in the other `~testOnly mypackage2.*`

Unfortunately, only one testOnly can run at a time. I appreciate this behaviour for compilation but not so much for other things which do not depend on each other. I guess sbt is conservative in this regard and assumes there is always some overlap between two test and executing both at the same time could lead to some race conditions/unexpected behaviour.

Maybe there is already a way to tell sbt client that executing two tasks of the same kind is allowed?

**Comment by adpi2** [comment]
Unfortunately it is not possible to run two tasks from two different clients in parallel.

However we have something called the `BackgroundJobService` that can be used in a task to start a job that will continue running after the task. This is used in `bgRun` to run the main class in the background, without blocking the task engine. We don't have a `bgTest` or `bgTestOnly` but we could implement it.

This is also something that could be contributed.

**Reply by robstoll** [reply]
Sounds not too bad. Say I implement bgTestOnly would it conflict with other jobs depending on compilation or are they still synchronised, i.e. only one runs and other waits for it to finish? What I don't want is that I run again in issues that two sbt taks compile at the same time and might conflict with each other.

**Reply by adpi2** [reply]
I think it will happen this way:
- the first task starts, compiles the code, copies the classes in some safe place and starts the tests in a new job
- the first task stops while the tests are still running
- the second task starts, it compiles the code (very fast because incremental so nothing to do), copies the classes and starts the tests in a new job
- it stops while the tests are still running
